### PR TITLE
Per user total walltime limit.

### PR DESCRIPTION
**Pull request by maxf130**
Add a limit type that makes it possible to limit a user to a total amount of wall time during a specified time window.  If (when a job is submitted) the total wall time for _finished_ jobs within the last n days (specified as `window` in the configuration) exceeds the specified limit, the job is paused.

### CODE REVIEW

File: config/job_conf.xml.sample_advanced [hunk 1]
```diff
@@ -762,6 +762,13 @@
                 will be terminated by Galaxy.
          -->
         <limit type="walltime">24:00:00</limit>
+	<!-- total_walltime:
+		Total walltime that jobs may not exceed during during a
```
Comment (mvdbeek) [inline config/job_conf.xml.sample_advanced]: s/during during/during/

File: config/job_conf.xml.sample_advanced [hunk 2]
```diff
@@ -762,6 +762,13 @@
...
+		set period.  If total walltime of finished jobs exceeds
+		this value, any new jobs are paused.  window is a number
```
Comment (mvdbeek) [inline config/job_conf.xml.sample_advanced]: s/  window/ Window/

File: lib/galaxy/jobs/handler.py [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import time
 import logging
 import threading
+import datetime
```
Comment (mvdbeek) [inline lib/galaxy/jobs/handler.py]: This should go to the top

**Review by mvdbeek — COMMENTED** [review]
This looks very good to me, thank you!

**Comment by maxf130** [comment]
Pleasure.  Is there any way of modifying the files within the pull request?  Or should I commit changes to dev and start a new one?

[EDIT]: Nevermind, I figured it out.  Pushing to [dev] did the trick.

File: config/job_conf.xml.sample_advanced (hunk 2 above)
Comment (maxf130) [reply config/job_conf.xml.sample_advanced]: window is referring to an attribute here, so it should probably be lowercase.  Maybe wrap in quotations marks?

**Comment by jmchilton** [comment]
@galaxybot test this

**Comment by maxf130** [comment]
I'm not entirely certain why the tests are failing.  Can I reproduce them locally? 

In principle I could check for the existance of `limits.total_walltime` before doing anything else, but fixing the test would be the better option.

**Comment by nsoranzo** [comment]
@maxf130 You need to add `total_walltime={}` also to the `Bunch()` at the end of `JobConfiguration.__parse_job_conf_legacy()` method in `lib/galaxy/jobs/__init__.py`

**Comment by maxf130** [comment]
@nsoranzo Thanks for the tip.

**Comment by nsoranzo** [comment]
@galaxybot test this

**Comment by jmchilton** [comment]
Thanks for the contribution. Looks pretty good at my first glance but I don't know much about the limits stuff. Ping @natefoo - any thoughts?

**Comment by natefoo** [comment]
LGTM. I think the total runtime-over-period calculation might be pretty expensive for busier sites, and I wonder whether the window ought to accept formats other than just days, but we can probably address those issues if they become problems.

Thanks @maxf130!

**Comment by bgruening** [comment]
@maxf130 great work! Thanks a lot! Do you have a twitter account? We would like to tweet your work and give you some credits!

**Comment by maxf130** [comment]
No problem!  Yes, my handle is @maxf130.
Thanks
Max

**Comment by bgruening** [comment]
https://twitter.com/bjoerngruening/status/805820789635293184
Thanks again @maxf130!

**Pull request merged**